gpcontrib/arenadata_toolkit/src/bloom_set.c

```diff
@@ -0,0 +1,280 @@
+/*
+ * Set of blooms. Main entry point to find a bloom and work with it.
+ * Used to track create, extend, truncate events.
+ */
+
+#include "bloom_set.h"
+
+#include <libpq/md5.h>
+
+#define BLOOM_ENTRY_GET(set, i) (void *)(set->bloom_entries + i * FULL_BLOOM_ENTRY_SIZE(set->bloom_size));
+
+static bloom_entry_t * bloom_entry_init(const uint32_t bloom_size, void *mem)
+{
+	bloom_entry_t *bloom_entry = mem;
+
+	bloom_entry->lock = LWLockAssign();
+	bloom_entry->dbid = InvalidOid;
+	(void)bloom_init(bloom_size, &bloom_entry->bloom);
+
+	return bloom_entry;
+}
+
+bloom_set_t *
+bloom_set_init(const uint32_t bloom_count, const uint32_t bloom_size, void *mem)
+{
+	bloom_set_t *bloom_set = mem;
+
+	bloom_set->bloom_count = bloom_count;
+	bloom_set->bloom_size = bloom_size;
+	bloom_set->lock = LWLockAssign();
+
+	for (uint32_t i = 0; i < bloom_count; i++)
+	{
+		void	   *bloom_entry_mem = BLOOM_ENTRY_GET(bloom_set, i);
+
+		(void)bloom_entry_init(bloom_size, bloom_entry_mem);
+	}
+
+	return bloom_set;
+}
+
+/* simple linear search, probably should be reworked (depends on target dbs count) */
+static bloom_entry_t * find_bloom_entry(bloom_set_t * bloom_set, Oid dbid, bool *found)
```

Comment:
We acquire the bloom_set lock each time we call this function. I think we can move lock and unlock inside this function.
Lock mode we can pass via the argument list. And also wether or not to lock the found entry.